### PR TITLE
ESQL: Document BUCKET as a grouping function

### PR DESCRIPTION
**Pull request by bpintea**
This adds the documentation for BUCKET as a grouping function and the addition of the "direct" invocation mode providing a span (in addition to the auto mode).

### CODE REVIEW

File: docs/reference/esql/functions/grouping-functions.asciidoc [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[[esql-group-functions]]
+==== {esql} aggregate functions
```
Comment (alex-spies) [inline docs/reference/esql/functions/grouping-functions.asciidoc:2]:
```suggestion
==== {esql} grouping functions
```

File: docs/reference/esql/functions/grouping-functions.asciidoc [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+++++
+<titleabbrev>Grouping functions</titleabbrev>
+++++
+
+The <<esql-stats-by>> command supports these grouping functions:
```
Comment (alex-spies) [inline docs/reference/esql/functions/grouping-functions.asciidoc:8]: Maybe we should add a short explanation?

E.g. "They are intended for use in the `BY` - clause, but can also be used in the aggregation expressions" (provided this is actually accurate :D )

File: docs/reference/esql/functions/examples/bucket.asciidoc [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 *Examples*
 
+`BUCKET` can work in two modes: one in which the size of the bucket is computed
+based on a buckets count recommendation (four parameters) and a range and
+another in which the bucket size is provided directly (two parameters).
+
```
Comment (alex-spies) [inline docs/reference/esql/functions/examples/bucket.asciidoc:8]: Should we also document that `BUCKET` has an alias, `BIN`?

File: docs/reference/esql/functions/examples/bucket.asciidoc [hunk 4]
```diff
@@ -58,6 +76,20 @@ include::{esql-specs}/bucket.csv-spec[tag=docsBucketNumeric-result]
 Unlike the earlier example that intentionally filters on a date range, you rarely want to filter on a numeric range.
 You have to find the `min` and `max` separately. {esql} doesn't yet have an easy way to do that automatically.
 
+If the desired bucket size is known in advance, simply provide it as the second
+argument, leaving the range out:
```
Comment (alex-spies) [inline docs/reference/esql/functions/examples/bucket.asciidoc:80]: nit: The repetition doesn't read so well.
```suggestion
As for dates, one can omit the range if the desired bucket size is known in advance. Simply provide it as the second
argument:
```

File: docs/reference/esql/functions/examples/bucket.asciidoc [hunk 5]
```diff
@@ -58,6 +76,20 @@ include::{esql-specs}/bucket.csv-spec[tag=docsBucketNumeric-result]
...
+----
+include::{esql-specs}/bucket.csv-spec[tag=docsBucketNumericWithSpan]
+----
+[%header.monospaced.styled,format=dsv,separator=|]
+|===
+include::{esql-specs}/bucket.csv-spec[tag=docsBucketNumericWithSpan-result]
+|===
+
+NOTE: When providing the bucket size as the second parameter, its type must be
+of a floating type.
```
Comment (alex-spies) [inline docs/reference/esql/functions/examples/bucket.asciidoc:91]: super nit:
```suggestion
NOTE: When providing the bucket size as the second parameter, its type must be
a floating point type.
```

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/bucket.csv-spec [hunk 6]
```diff
@@ -205,6 +205,27 @@ FROM employees
...
+;
+
+//tag::docsBucketWeeklyHistogramWithSpan-result[]
+  hires_per_week:long    |    week:date
+2              |1985-02-18T00:00:00.000Z
+1              |1985-05-13T00:00:00.000Z
+1              |1985-07-08T00:00:00.000Z
+1              |1985-09-16T00:00:00.000Z
+2              |1985-10-14T00:00:00.000Z
+4              |1985-11-18T00:00:00.000Z
```
Comment (alex-spies) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/bucket.csv-spec:225]: nit: formatting

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java [hunk 7]
```diff
@@ -92,6 +92,10 @@ public class Bucket extends GroupingFunction implements Validatable, TwoOptional
         examples = {
             @Example(
                 description = """
+                    `BUCKET` can work in two modes: one in which the size of the bucket is computed
+                    based on a buckets count recommendation (four parameters) and a range and
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java:96]: nit: hard to grok the long sentence, maybe a comma would help?
```suggestion
                    based on a buckets count recommendation (four parameters) and a range, and
```

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java [hunk 8]
```diff
@@ -120,6 +124,11 @@ public class Bucket extends GroupingFunction implements Validatable, TwoOptional
                     For rows with a value outside of the range, it returns a bucket value that corresponds to a bucket outside the range.
                     Combine`BUCKET` with <<esql-where>> to filter rows."""
             ),
+            @Example(description = """
+                If the desired bucket size is known in advance, simply provide it as the second
+                argument, leaving the range out:""", file = "bucket", tag = "docsBucketWeeklyHistogramWithSpan", explanation = """
+                NOTE: When providing the bucket size as the second parameter, its type must be
+                of a time duration or date period type."""),
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java:131]: super nit:
```suggestion
                NOTE: When providing the bucket size as the second parameter, it must be
                a time duration or date period."""),
```

**Review by alex-spies — COMMENTED** [review]
I think we should update the parameter descriptions as well, to mention that `from` and `to` can be omitted (together):
image

File: docs/reference/esql/functions/grouping-functions.asciidoc (hunk 1 above)
Comment (bpintea) [reply docs/reference/esql/functions/grouping-functions.asciidoc]: Thanks, fixed.

File: docs/reference/esql/functions/grouping-functions.asciidoc (hunk 2 above)
Comment (bpintea) [reply docs/reference/esql/functions/grouping-functions.asciidoc:8]: Good point -- I've added an explanation with a BUCKET example. If we'll add another function, we can generalise. We don't say much for aggregate functions either (where one can use them and how), so I guess examples help better?

File: docs/reference/esql/functions/examples/bucket.asciidoc (hunk 4 above)
Comment (bpintea) [reply docs/reference/esql/functions/examples/bucket.asciidoc]: Applied (left out the dates reference, though).

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/bucket.csv-spec (hunk 6 above)
Comment (bpintea) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/bucket.csv-spec]: Reformatted (all around).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java (hunk 7 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java:96]: Added.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java (hunk 8 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java:131]: Rephrased.

**Review by alex-spies — APPROVED** [review]

**Comment by bpintea** [comment]
Thanks, Alex!

**Pull request merged**